### PR TITLE
feat: GCI-15 implement step

### PR DESCRIPTION
**Pull request by MrBlaise**
### Checklist

- [ ] I've read and followed the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- [ ] `step.yml` and `README.md` is updated with the changes (if needed)

### Version

Requires a *MAJOR/MINOR/PATCH* [version update](https://semver.org/)

### CODE REVIEW

File: step/step.go [hunk 1]
```diff
@@ -42,21 +67,221 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+
+	switch {
+	case input.UseBitriseCache:
+		args = append(args, fmt.Sprintf("--cache-from=type=local,src=%s", dockerCacheFolder))
+		args = append(args, fmt.Sprintf("--cache-to=type=local,dest=%s,mode=max,compression=zstd", dockerCacheFolderTemporary))
+	case input.CacheFrom != "":
+		for _, cacheFrom := range strings.Split(input.CacheFrom, "\n") {
+			args = append(args, fmt.Sprintf("--cache-from=%s", cacheFrom))
+		}
+		fallthrough
```
Comment (MrBlaise) [inline step/step.go:194]: This fallthrough could look weird, and we apologize for it, but thinking through all the alternative implementations funnily enough this seemed the less complex (to us) :D 

Logic: 
- When UseBitriseCache is true, we only want to set our fix cache from and cache-to params
- Otherwise we want to set cache-from and cache-to individually only and only if they are not empty (because adding an empty value to the array will break the cmdrunner)

File: .yamllint.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+extends: default
```
Comment (ofalvai) [inline .yamllint.yml]: This file is not necessary, we have a global lint config, enforced for all steps: https://github.com/bitrise-steplib/steps-check/blob/master/.yamllint.yml

File: step.yml [hunk 3]
```diff
@@ -104,6 +123,62 @@ inputs:
       value_options:
         - "true"
         - "false"
+      is_required: true
+
+  - build_arg:
+    opts:
+      title: Build arguments
+      summary: List of build arguments to be passed to the docker build
```
Comment (ofalvai) [inline step.yml]:
```suggestion
      summary: List of [build arguments](https://docs.docker.com/build/guide/build-args/) to be passed to the docker build
```
Let's save a Google search for users and add links to the official docs wherever possible

File: step/step.go [hunk 4]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+}
+
+func (step DockerBuildPushStep) restoreCache(input Input, imageName string) error {
+	step.logger.Infof("Restoring cache...")
+	saver := cache.NewRestorer(step.envRepo, step.logger, step.commandFactory)
+
+	var cacheKey = []string{
+		fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}-{{ .Branch }}", imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}", imageName),
```
Comment (ofalvai) [inline step/step.go:106]: Listing these 3 keys tell me that maybe `imageName` should come before the other two variables and then a single cache key would be enough. Or did I miss something?

File: step/step.go [hunk 5]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+		if err := step.saveCache(input, tagUsedInCacheKey); err != nil {
+			return fmt.Errorf("save cache: %w", err)
+		}
+	}
+	return nil
+}
+
+func (step DockerBuildPushStep) restoreCache(input Input, imageName string) error {
+	step.logger.Infof("Restoring cache...")
+	saver := cache.NewRestorer(step.envRepo, step.logger, step.commandFactory)
```
Comment (ofalvai) [inline step/step.go]: nit: You probably forgot to rename the variable 😄

File: step/step.go [hunk 6]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+
+	var cacheKey = []string{
+		fmt.Sprintf(dockerCacheKeyTemplate, imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}-{{ .Branch }}", imageName),
+		fmt.Sprintf("docker-%s-{{ .OS }}-{{ .Arch }}", imageName),
+	}
+
+	return saver.Restore(cache.RestoreCacheInput{
+		StepId:  "restore-cache",
```
Comment (ofalvai) [inline step/step.go]: Please update the step ID, this is part of the telemetry events to see which step invoked the shared lib

File: step/step.go [hunk 7]
```diff
@@ -27,7 +38,21 @@ type DockerBuildPushStep struct {
 	envRepo        env.Repository
 }
 
-func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory command.Factory, pathChecker pathutil.PathChecker, pathProvider pathutil.PathProvider, pathModifier pathutil.PathModifier, envRepo env.Repository) DockerBuildPushStep {
+const (
+	dockerCacheKeyTemplate     = "docker-%s-{{ .OS }}-{{ .Arch }}-{{ .Branch }}-{{ .CommitHash }}"
```
Comment (ofalvai) [inline step/step.go:43]: Shouldn't a checksum be part of the key (probably the checksum of the Dockerfile)?

File: step/step.go [hunk 8]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+		}
+	}()
+
+	if err := step.build(input); err != nil {
+		return fmt.Errorf("build docker image: %w", err)
+	}
+
+	// To make sure the cache is not growing indefinitely,
+	// we remove the original cache folder and move the new one to its place
+	if err := step.moveCacheFolder(dockerCacheFolderTemporary, dockerCacheFolder); err != nil {
```
Comment (ofalvai) [inline step/step.go:155]: Did you look into the `--filter` flag of `buildx prune`? https://docs.docker.com/engine/reference/commandline/buildx_prune/

Sounds like we could do something similar with recording the timestamp of starting the build, then clean up everything that was created before that timestamp? Some steps have this exact mechanism for dealing with files restored from cache vs. files produced during the build.

File: step/step.go (hunk 1 above)
Comment (ofalvai) [reply step/step.go:194]: This sounds reasonable to me, but then it should be documented that the cache inputs and the Bitrise cache input are mutually exclusive

File: step/step.go [hunk 9]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+		for _, cacheTo := range strings.Split(input.CacheTo, "\n") {
+			args = append(args, fmt.Sprintf("--cache-to=%s", cacheTo))
+		}
+	}
+
+	if input.ExtraOptions != "" {
+		for _, option := range strings.Split(input.ExtraOptions, "\n") {
+			// This regex splits the string by spaces, but keeps quoted strings together
+			// Example --build-arg "-X main.version=1.0.0" will be split into --build-arg and "-X main.version=1.0.0"
+			r := regexp.MustCompile(`[^\s"']+|"([^"]*)"|'([^']*)`)
```
Comment (ofalvai) [inline step/step.go]: I'm not against (all) regex, but sounds like this stdlib package could do something like what we need: https://pkg.go.dev/flag

File: step/step.go [hunk 10]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+
+	for _, tag := range strings.Split(input.Tags, "\n") {
+		args = append(args, "--tag", tag)
+	}
+
+	args = append(args, []string{"-f", input.File, input.Context}...)
+
+	step.logger.Infof("$ docker %s", strings.Join(args, " "))
+
+	buildxCmd := step.commandFactory.Create("docker", args, &command.Opts{
```
Comment (ofalvai) [inline step/step.go:225]: You can just pass `nil` as the `Opts` if you don't want to customize the default behavior (which connects stdout to stdout, stderr to stderr).

File: step/step.go [hunk 11]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+		args = append(args, "--driver-opt", "network=host", "--buildkitd-flags", "--allow-insecure-entitlement network.host")
+	}
+
+	createCmd := step.commandFactory.Create("docker", args, nil)
+
+	step.logger.Infof("$ docker %s", strings.Join(args, " "))
+
+	out, err := createCmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("create buildx instance: %w", err)
```
Comment (ofalvai) [inline step/step.go]: This won't print the contents of `stderr` when you use `...CombinedOutput()`, `err` contains only something like `exit status 2`. We usually print both `err` and `out` in this case.

File: step/step.go [hunk 12]
```diff
@@ -42,21 +67,227 @@ func New(logger log.Logger, inputParser stepconf.InputParser, commandFactory com
...
+	}
+	return out, nil
+}
+
+func (step DockerBuildPushStep) createCacheFolder(path string) error {
+	cmd := step.commandFactory.Create("mkdir", []string{"-p", path}, nil)
```
Comment (ofalvai) [inline step/step.go]: Is there any reason you used shell commands instead of the `io` or `ioutils` package?

File: step/util.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
+package step
+
+type InfoLogger interface {
+	Infof(format string, v ...interface{})
+}
+
+// LogWriter can be used in place of an io.Writer where we want to redirect the output to a logger in real time
+type LogWriter struct {
```
Comment (ofalvai) [inline step/util.go]: I'm not sure I understand why it's needed at all. Isn't it enough to use stdout and stderr when running the Docker commands?

File: .yamllint.yml (hunk 2 above)
Comment (MrBlaise) [reply .yamllint.yml]: Without this the liner was very unreasonable (no empty lines at all, and enforcing array indentation), at least for me. Do I need to set something up globally first for it to work?

File: step/step.go (hunk 7 above)
Comment (MrBlaise) [reply step/step.go:43]: With dockerfiles that does not work reliably as the image changes often based on the content not the dockerfile. So if any of the code changes and you have COPY command that would result in a new image.

File: step/step.go (hunk 8 above)
Comment (MrBlaise) [reply step/step.go:155]: We need the old cache during the build so pruneing before is not an option, pruning after using the time could be risky we could accidentally remove less or more. This solution has less variables and guaranteed to only cache exactly what we need.

File: .yamllint.yml (hunk 2 above)
Comment (ofalvai) [reply .yamllint.yml]: No need to set up anything at all, just run `bitrise run check`.

Comment (ofalvai) [reply .yamllint.yml]: Oh I noticed the `check` and `e2e` workflows are not exactly as they are in other steps: https://github.com/bitrise-steplib/bitrise-step-restore-cache/blob/main/bitrise.yml#L5

That `check` step is a bit messy and undocumented, sorry about that 🙈 Both workflows should run the check step, but if you don't change the default step inputs, it runs everything except for E2E tests.

File: step/step.go (hunk 1 above)
Comment (MrBlaise) [reply step/step.go:194]: It is documented in the step.yml which should show up on the UI as well right?

File: .yamllint.yml (hunk 2 above)
Comment (MrBlaise) [reply .yamllint.yml]: Yeah that kept failing for me it was super strict until I added this, we could try it when we sync about this PR.

File: step/step.go (hunk 7 above)
Comment (ofalvai) [reply step/step.go:43]: Yeah, makes sense. We would probably need to checksum every single file inside the context folder, but this seems a bit overkill.

File: step/step.go (hunk 8 above)
Comment (ofalvai) [reply step/step.go:155]: I'm okay with it, let's go this way first. We could iterate on it in the future

File: step/step.go (hunk 4 above)
Comment (MrBlaise) [reply step/step.go:106]: What would the exact cache key look like in that case I am not sure I understand.

The reasoning for this here is to have a fallback hierarchy like so:

If we have built the same image already -> use that
otherwise check if we have built it already in the same branch
otherwise check if we have ever built it for this particular os-arch
otherwise this is the first time we are seeing it.

File: step/util.go (hunk 13 above)
Comment (MrBlaise) [reply step/util.go]: That would only output the logs after the build has finished which could be a while. This is both bad UX and I believe there is a timeout for steps that might kill it this way.

If you mean the actual os stdout, that would not go through the step logger so it will not be the same structure making it probably look out of place, or hard to parse.

Comment (bvatai-br) [reply step/util.go]: simply running and awaiting cmd completion: execution is running in the background, no output is written to stdout, and when the process has returned, everything is flushed at one big take.  This is not a nice UX, also I'm afraid it'd cause false timeouts in edge cases. Instead, with this implementation we yield output line-by-line, as it is coming from docker.

File: step/step.go (hunk 9 above)
Comment (MrBlaise) [reply step/step.go]: I am not sure this applies here. Flag is used for parsing predefined/known flag from the os.Args of the go program.

Here we accept a string variable through the step input which we need to then pass to the docker command. So this won't help in parsing the string, nor would we want to define all the possible docker commands here.

File: step/step.go (hunk 10 above)
Comment (MrBlaise) [reply step/step.go:225]: We do want to customize, this stdout variable references our custom logwriter

File: step/step.go (hunk 12 above)
Comment (MrBlaise) [reply step/step.go]: No reason, we just mapped the POC from a script directly and forgot to think about to go native solution for this

Comment (MrBlaise) [reply step/step.go]: Fixed

File: step/step.go (hunk 10 above)
Comment (MrBlaise) [reply step/step.go:225]: We removed the logger wrapper as stdout works too to reduce complexity.

**Review by ofalvai — DISMISSED** [review]

**Review by bvatai-br — APPROVED** [review]
